bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java

```diff
@@ -1947,6 +1937,156 @@ protected void destroy() {
+		}
+		ElementAtZoom<ImageData> loadedImageData = loadImageData(zoom);
+		return DPIUtil.scaleImageData(device, loadedImageData, zoom);
+	}
+
+	@Override
+	ImageHandle getImageMetadata(int zoom) {
+		if (zoomLevelToImageHandle.containsKey(zoom)) {
+			return zoomLevelToImageHandle.get(zoom);
+		} else  {
+			ImageData scaledImageData = getImageData(zoom);
+			ImageHandle imageHandle = init(scaledImageData, zoom);
+			return imageHandle;
+		}
+	}
+}
+
+private class PlainImageDataProviderWrapper extends ImageFromImageDataProviderWrapper {
+	private ImageData imageDataAt100;
+
+	PlainImageDataProviderWrapper(ImageData imageData) {
+		this.imageDataAt100 = (ImageData) imageData.clone();
+		initImage();
+	}
+
+	@Override
+	protected Rectangle getBounds(int zoom) {
+		Rectangle rectangle = new Rectangle(0, 0, imageDataAt100.width, imageDataAt100.height);
+		return DPIUtil.scaleUp(rectangle, zoom);
+	}
+
+	@Override
+	protected ElementAtZoom<ImageData> loadImageData(int zoom) {
+		return new ElementAtZoom<>(imageDataAt100, 100);
+	}
+
+	@Override
+	AbstractImageProviderWrapper createCopy(Image image) {
+		return image.new PlainImageDataProviderWrapper(this.imageDataAt100);
+	}
+}
+
+private class MaskedImageDataProviderWrapper extends ImageFromImageDataProviderWrapper {
+	private final ImageData srcAt100;
+	private final ImageData maskAt100;
+
+	MaskedImageDataProviderWrapper(ImageData srcAt100, ImageData maskAt100) {
+		this.srcAt100 = (ImageData) srcAt100.clone();
+		this.maskAt100 = (ImageData) maskAt100.clone();
+		initImage();
+	}
+
+	@Override
+	protected Rectangle getBounds(int zoom) {
+		Rectangle rectangle = new Rectangle(0, 0, srcAt100.width, srcAt100.height);
+		return DPIUtil.scaleUp(rectangle, zoom);
+	}
+
+	@Override
+	protected ElementAtZoom<ImageData> loadImageData(int zoom) {
+		ImageData scaledSource = DPIUtil.scaleImageData(device, srcAt100, zoom, 100);
+		ImageData mask = ImageData.convertMask(maskAt100);
+		ImageData scaledMask = DPIUtil.scaleImageData(device, mask, zoom, 100);
+		ImageData mergedData = applyMask(scaledSource, scaledMask);
+		return new ElementAtZoom<>(mergedData, zoom);
+	}
+
+	@Override
+	AbstractImageProviderWrapper createCopy(Image image) {
+		return image.new MaskedImageDataProviderWrapper(this.srcAt100, this.maskAt100);
+	}
+}
+
+private class ImageDataLoaderStreamProviderWrapper extends ImageFromImageDataProviderWrapper {
+	private byte[] inputStreamData;
+
+	ImageDataLoaderStreamProviderWrapper(InputStream inputStream) {
+		try {
+			this.inputStreamData = inputStream.readAllBytes();
+			initImage();
+		} catch (IOException e) {
+			SWT.error(SWT.ERROR_INVALID_ARGUMENT, e);
+		}
+	}
+
+	private ImageDataLoaderStreamProviderWrapper(byte[] inputStreamData) {
+		this.inputStreamData = inputStreamData;
+	}
+
+	@Override
+	protected ElementAtZoom<ImageData> loadImageData(int zoom) {
+		return ImageDataLoader.load(new ByteArrayInputStream(inputStreamData), FileFormat.DEFAULT_ZOOM, zoom);
```

Comment:
This is a nice improvement as in case you have a stream for a dynamically-sized image (like an SVG) you can then create it at the appropriate zoom.